src/extractor.rs

```diff
@@ -418,30 +468,57 @@ fn type_check(ast: ExprU) -> Result<ExprT, TypeError> {
                 "ref" => {
-                    if args.len() != 1 && args.len() != 2 {
-                        return Err(TypeError::ArgumentMismatch {
-                            expected: vec![1, 2],
-                            found: args.len(),
-                        });
-                    }
-                    let typed_args = args
-                        .into_iter()
-                        .map(|arg| {
-                            match arg {
-                                // refs can only take string literals
-                                ExprU::StringU(s) => Ok(s),
-                                // error on everything that isn't a string literal
-                                not_string => Err(TypeError::TypeMismatch {
-                                    expected: ExprType::String,
-                                    got: ExprType::from(&not_string),
-                                }),
-                            }
-                        })
-                        .collect::<Result<Vec<String>, TypeError>>()?;
-                    Ok(ExprT::RefT(
-                        typed_args[0].clone(),
-                        typed_args.get(1).map(|x| x.to_owned()),
-                    ))
+                    let mut arg_stack = Vec::from(args);
+                    arg_stack.reverse();
+
+                    // The first param must be there, and must be a string.
+                    let p1: String = match arg_stack.pop() {
+                        None => Err(TypeError::ArgumentMismatch {
+                            expected: vec![1, 2, 3],
+                            found: 0,
+                        }),
+                        Some(ExprU::StringU(s)) => Ok(s.clone()),
+                        Some(a) => Err(TypeError::TypeMismatch {
+                            expected: ExprType::String,
+                            got: ExprType::from(&a),
+                        }),
+                    }?;
+
+                    // The second param is optional and might be a string.
+                    let p2: Option<String> = match arg_stack.pop() {
+                        None => None,
+                        Some(ExprU::StringU(s)) => Some(s.clone()),
+                        Some(x) => {
+                            arg_stack.push(x);
+                            None
+                        }
+                    };
+
+                    // If there are two string params, the first is the package
+                    // and the second is the name. If there is one, it's just the
+                    // name. This little match sorts that out.
+                    let (name, package) = match (p1, p2) {
+                        (a, Some(b)) => (b, Some(a)),
+                        (a, None) => (a, None)
+                    };
+
+                    // Now deal with additional params. There should be at most
+                    // one, and it must be a kwarg with key 'v' or 'version'.
```

Comment:
If there should be at most one, we should error if there's more than one.

Reply:
Now handled.